import pandas as pd
import numpy as np

In [2]:
import pandas as pd

url = "https://www.sharkattackfile.net/spreadsheets/GSAF5.xls"

df_shark_attacks = pd.read_excel(url)

df_shark_attacks

Date    Year        Type             Country  \
0      10th January  2026.0  Unprovoked           Australia   
1       8th January  2026.0  Unprovoked   US Virgin Islands   
2      3rd January   2026.0  Unprovoked       New Caledonia   
3     21st December  2025.0  Unprovoked                 USA   
4     12th December  2025.0  Unprovoked                 USA   
...             ...     ...         ...                 ...   
7060    Before 1903     0.0  Unprovoked           AUSTRALIA   
7061    Before 1903     0.0  Unprovoked           AUSTRALIA   
7062      1900-1905     0.0  Unprovoked                 USA   
7063      1883-1889     0.0  Unprovoked              PANAMA   
7064      1845-1853     0.0  Unprovoked  CEYLON (SRI LANKA)   

                             State                             Location  \
0                              NSW                         Avalon Beach   
1     Fredricksted Island St Croix                         Dorsch Beach   
2                             Kélé          Between Bourail and Moindou   
3                       California           Lovers Point Pacific Grove   
4         Sonoma County California                         Salmon Creek   
...                            ...                                  ...   
7060             Western Australia                          Roebuck Bay   
7061             Western Australia                                  NaN   
7062                North Carolina                       Ocracoke Inlet   
7063                           NaN                 Panama Bay 8ºN, 79ºW   
7064              Eastern Province  Below the English fort, Trincomalee   

          Activity                   Name Sex  Age  ...  \
0          Surfing           Paul Stanton   M    ?  ...   
1       Snorkeling          Arlene Lillis   F   56  ...   
2     Scuba Diving                Unknown   M    ?  ...   
3         Swimming              Erica Fox   F   55  ...   
4          Surfing                Unknown   M    ?  ...   
...            ...                    ...  ..  ...  ...   
7060        Diving                   male   M  NaN  ...   
7061  Pearl diving                  Ahmun   M  NaN  ...   
7062      Swimming  Coast Guard personnel   M  NaN  ...   
7063           NaN        Jules Patterson   M  NaN  ...   
7064      Swimming                   male   M   15  ...   

                         Species   \
0                         Unknown   
1                        Unknown    
2                         Unknown   
3               Great White Shark   
4     Suspected Great White Shark   
...                           ...   
7060                          NaN   
7061                          NaN   
7062                          NaN   
7063                          NaN   
7064                          NaN   

                                                 Source  \
0                                        Bob Myatt GSAF   
1          Todd Smith: KevinMcMurray Trackingsharks.com   
2                            Andy Currie: Province Sud:   
3     Kevin McMurray Tracking sharks.com: Ralph Coll...   
4     Kevin McMurray Tracking sharks.com:Andrew Curr...   
...                                                 ...   
7060                   H. Taunton; N. Bartlett,  p. 234   
7061              H. Taunton; N. Bartlett,  pp. 233-234   
7062               F. Schwartz, p.23; C. Creswell, GSAF   
7063                                The Sun, 10/20/1938   
7064                                         S.W. Baker   

                                 pdf  \
0                                NaN   
1                                NaN   
2                                NaN   
3                                NaN   
4                                NaN   
...                              ...   
7060          ND-0005-RoebuckBay.pdf   
7061               ND-0004-Ahmun.pdf   
7062  ND-0003-Ocracoke_1900-1905.pdf   
7063      ND-0002-JulesPatterson.pdf   
7064              ND-0001-Ceylon.pdf   

             

In [3]:
print(df_shark_attacks.shape)

(7065, 23)


# Conclusiones de la exploración inicial

1. El dataset tiene `7065` filas y `23` columnas. (`(7065, 23)`)

In [4]:
columnas_numericas = df_shark_attacks.select_dtypes(include=['number']).columns.tolist()
print (f"Las columnas numericas son : {columnas_numericas}")
categorical_columns = df_shark_attacks.select_dtypes(include=['object']).columns.tolist()
print(f"Las columnas categoricas serian: {categorical_columns}")

Las columnas numericas son : ['Year', 'original order']
Las columnas categoricas serian: ['Date', 'Type', 'Country', 'State', 'Location', 'Activity', 'Name', 'Sex', 'Age', 'Injury', 'Fatal Y/N', 'Time', 'Species ', 'Source', 'pdf', 'href formula', 'href', 'Case Number', 'Case Number.1', 'Unnamed: 21', 'Unnamed: 22']


2. Hay varios tipos de datos que no son apropiados para lo que describen:
    - `Year` aparece como `float64`, pero conceptualmente es un año (debería ser un `int` o tratarse como temporal/categórico según el análisis).
    - `Age` aparece como `object`, pero es **numérico** (requiere limpieza).
    - `Date` y `Time` aparecen como `object`, pero son variables **temporales**.
    - `Fatal Y/N` y `Sex` aparecen como `object`, y deberían formatearse y limpiarse.

In [5]:
df_shark_attacks.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7065 entries, 0 to 7064
Data columns (total 23 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Date            7065 non-null   object 
 1   Year            7063 non-null   float64
 2   Type            7047 non-null   object 
 3   Country         7015 non-null   object 
 4   State           6578 non-null   object 
 5   Location        6498 non-null   object 
 6   Activity        6480 non-null   object 
 7   Name            6846 non-null   object 
 8   Sex             6486 non-null   object 
 9   Age             4070 non-null   object 
 10  Injury          7030 non-null   object 
 11  Fatal Y/N       6504 non-null   object 
 12  Time            3538 non-null   object 
 13  Species         3934 non-null   object 
 14  Source          7045 non-null   object 
 15  pdf             6799 non-null   object 
 16  href formula    6794 non-null   object 
 17  href            6796 non-null   o

In [5]:
porcentage_null = df_shark_attacks.isnull().sum()/ len(df_shark_attacks)*100
porcentage_null

Date               0.000000
Year               0.028309
Type               0.254777
Country            0.707714
State              6.893135
Location           8.025478
Activity           8.280255
Name               3.099788
Sex                8.195329
Age               42.392074
Injury             0.495400
Fatal Y/N          7.940552
Time              49.922151
Species           44.317056
Source             0.283086
pdf                3.765039
href formula       3.835810
href               3.807502
Case Number        3.779193
Case Number.1      3.793347
original order     3.765039
Unnamed: 21       99.985846
Unnamed: 22       99.971691
dtype: float64

# Porcentaje de valores nulos en cada Columna:

- La mayoria de columnas tiene menos del 10% de valores nulos:
- `Age`, `Time`, `Species` y `Unnamed: 21` y `Unnamed: 22` tienen 42%, 49%, 44%, 99% y 99% respectivamente.

Estos porcentajes hacen necesaria la valoración de si son necesarios para el análisis.

# Valores únicos de cada columna determinando las que podrían ser categóricas / temporales / identificadores:

In [8]:
df_shark_attacks['Year'].dtype



dtype('float64')

In [11]:
print( df_shark_attacks["Year"].unique())

[2026. 2025. 2024. 2023. 2022. 2021. 2020. 2019. 2018. 2017.   nan 2016.
 2015. 2014. 2013. 2012. 2011. 2010. 2009. 2008. 2007. 2006. 2005. 2004.
 2003. 2002. 2001. 2000. 1999. 1998. 1997. 1996. 1995. 1984. 1994. 1993.
 1992. 1991. 1990. 1989. 1969. 1988. 1987. 1986. 1985. 1983. 1982. 1981.
 1980. 1979. 1978. 1977. 1976. 1975. 1974. 1973. 1972. 1971. 1970. 1968.
 1967. 1966. 1965. 1964. 1963. 1962. 1961. 1960. 1959. 1958. 1957. 1956.
 1955. 1954. 1953. 1952. 1951. 1950. 1949. 1948. 1848. 1947. 1946. 1945.
 1944. 1943. 1942. 1941. 1940. 1939. 1938. 1937. 1936. 1935. 1934. 1933.
 1932. 1931. 1930. 1929. 1928. 1927. 1926. 1925. 1924. 1923. 1922. 1921.
 1920. 1919. 1918. 1917. 1916. 1915. 1914. 1913. 1912. 1911. 1910. 1909.
 1908. 1907. 1906. 1905. 1904. 1903. 1902. 1901. 1900. 1899. 1898. 1897.
 1896. 1895. 1894. 1893. 1892. 1891. 1890. 1889. 1888. 1887. 1886. 1885.
 1884. 1883. 1882. 1881. 1880. 1879. 1878. 1877. 1876. 1875. 1874. 1873.
 1872. 1871. 1870. 1869. 1868. 1867. 1866. 1865. 18

Comprobacion de valores unicos en columna Year

In [10]:

print(df_shark_attacks['Year'].isnull().sum())

2
